### PR TITLE
Make some of lexer's API private

### PR DESCRIPTION
**Pull request by matklad**
Lexer is a `pub` type, so it feels wrong that its fields are just pub (I guess it wasn't exported initially), so let's minimize visibility.

Context: I am looking into extracting rust-lexer into a library, which can be shared by rust-analyzer and rustc. I hope that a simple interface like `fn next_token(src: &str) -> (TokenKind, usize)` would work, but to try this out I need to understand what is the current API of the lexer.

### CODE REVIEW

File: src/libsyntax/parse/lexer/mod.rs [hunk 1]
```diff
@@ -43,16 +43,16 @@ pub struct UnmatchedBrace {
 }
 
 pub struct StringReader<'a> {
-    pub sess: &'a ParseSess,
+    crate sess: &'a ParseSess,
     /// The absolute offset within the source_map of the next character to read
-    pub next_pos: BytePos,
+    crate next_pos: BytePos,
```
Comment (lzutao) [inline src/libsyntax/parse/lexer/mod.rs]: A bikeshed question: What does `crate` keyword do here? I cannot find the answer in [rust reference book](https://doc.rust-lang.org/nightly/reference/visibility-and-privacy.html).

Comment (matklad) [reply src/libsyntax/parse/lexer/mod.rs]: It's an unstable syntax for `pub(crate)` visibility

Comment (lzutao) [reply src/libsyntax/parse/lexer/mod.rs]: Thank you.

File: src/libsyntax/parse/lexer/mod.rs [hunk 2]
```diff
@@ -1911,7 +1911,7 @@ mod tests {
 
     use crate::ast::{Ident, CrateConfig};
     use crate::symbol::Symbol;
-    use crate::source_map::SourceMap;
+    use crate::source_map::{SourceMap, FilePathMapping};
```
Comment (sanxiyn) [inline src/libsyntax/parse/lexer/mod.rs]: This seems included by mistake.

Comment (matklad) [reply src/libsyntax/parse/lexer/mod.rs]: :+1: that was indeed intended for https://github.com/rust-lang/rust/pull/59687

**Comment by eddyb** [comment]
> I hope that a simple interface like `fn next_token(src: &str) -> (TokenKind, usize)` would work, but to try this out I need to understand what is the current API of the lexer.

The API feels like it predates iterators, in some sense, it should probably be streamlined.

**Comment by eddyb** [comment]
For example, the peeking should ideally be handled with `.peekable()`.

Anyway, this seems like a good start.

@bors r+

**Comment by matklad** [comment]
@eddyb yeah, I feel like this could be improved **massively** by separating the actual lexing from rustc-specific bits like spans and source maps.  See https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/Extracting.20the.20lexer/near/162515922 for the rough proposal

**Comment by bjorn3** [comment]
Why is bors testing this and #59556? It is already merged.

**Pull request merged**